common/metrics/grpc.go

```diff
@@ -0,0 +1,156 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"context"
+
+	metricspb "go.temporal.io/server/api/metrics/v1"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type metricsContextKey struct{}
+
+var (
+	baggageTrailerKey = "metrics-baggage-bin"
+	metricsCtxKey     = metricsContextKey{}
+)
+
+// NewServerMetricsContextInjectorInterceptor returns grpc server interceptor that wraps golang context into golang
+// metrics propagation context.
+func NewServerMetricsContextInjectorInterceptor() grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		ctxWithMetricsBaggage := AddMetricsBaggageToContext(ctx)
+		return handler(ctxWithMetricsBaggage, req)
+	}
+}
+
+// NewClientMetricsTrailerPropagatorInterceptor returns grpc client interceptor that injects metrics propagation context
+// received in trailer into golang metrics propagation context.
+func NewClientMetricsTrailerPropagatorInterceptor(logger log.Logger) grpc.UnaryClientInterceptor {
+	return func(
+		ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker,
+		opts ...grpc.CallOption,
+	) error {
+		var trailer metadata.MD
+		optsWithTrailer := append(opts, grpc.Trailer(&trailer))
+		err := invoker(ctx, method, req, reply, cc, optsWithTrailer...)
+
+		metricsBaggageStrings := trailer.Get(baggageTrailerKey)
+		if metricsBaggageStrings == nil {
```

Comment:
```suggestion
		if len(metricsBaggageStrings) == 0 {
```
covers both `nil` and zero length.